File: 20.有效的括号.cs
/*
 * @lc app=leetcode.cn id=20 lang=csharp
 *
 * [20] 有效的括号
 */

// @lc code=start
public class Solution
{
    public bool IsValid(string s)
    {
        int len = s.Length;
        if (len % 2 != 0)
        {
            return false;
        }
        Stack<char> chars = new();
        for (int i = 0; i < len; i++)
        {
            char c = s[i];
            if (c == '(' || c == '[' || c == '{')
            {
                chars.Push(c);
            }
            else if (chars.Count == 0 || (c == ')' && chars.Pop() != '(' || (c == ']' && chars.Pop() != '[') || (c == '}' && chars.Pop() != '{')))
            {
                return false;
            }
        }
        return chars.Count == 0;
    }
}
// @lc code=end

File: 20.有效的括号.rs
/*
 * @lc app=leetcode.cn id=20 lang=rust
 *
 * [20] 有效的括号
 */

// @lc code=start
impl Solution {
    pub fn is_valid(s: String) -> bool {
        if s.len() % 2 == 1 {
            return false;
        }
        let mut stack: Vec<char> = vec![];
        for c in s.chars() {
            match c {
                '(' | '[' | '{' => stack.push(c),
                ')' => {
                    if stack.len() == 0 || stack.pop().unwrap() != '(' {
                        return false;
                    }
                }
                ']' => {
                    if stack.len() == 0 || stack.pop().unwrap() != '[' {
                        return false;
                    }
                }
                '}' => {
                    if stack.len() == 0 || stack.pop().unwrap() != '{' {
                        return false;
                    }
                }
                _ => (),
            }
        }
        stack.len() == 0
    }
}
// @lc code=end

